### PR TITLE
[IPU] Add reset dataloader hooks to training type plugin 3/n

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

This PR adds reset data-loader hooks to the training type plugin and accelerator.

The requirement here was that `process_dataloader` (The currently exposed hook) happens too late in the flow, not updating the progress bar etc. This till now has been fine as the assumption is the size of the batches do not change, however this is not true with IPUs.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/accelerators/accelerator.py [hunk 1]
```diff
@@ -460,6 +460,22 @@ def process_dataloader(self, dataloader: Union[Iterable, DataLoader]) -> Union[I
         """
         return self.training_type_plugin.process_dataloader(dataloader)
 
+    def on_reset_train_dataloader(self, dataloader) -> Union[Iterable, DataLoader]:
```
Comment (awaelchli) [inline pytorch_lightning/accelerators/accelerator.py:463]: the annotation for dataloader is missing :)

File: tests/accelerators/test_cpu.py [hunk 2]
```diff
@@ -50,3 +51,97 @@ def setup_optimizers_in_pre_dispatch(self) -> bool:
...
+    trainer.validate(model)
+    trainer.test(model)
+    trainer.predict(model)
+    # assert val/test/predict loader hooks were called
+    assert accelerator.mock_count == 3
+
+
+def test_plugin_on_reset_dataloader_hooks(tmpdir):
+    """
+    Ensures data-loader hooks are called at the right stages when using a plugin.
```
Comment (awaelchli) [inline tests/accelerators/test_cpu.py:105]: ok it doesn't really do that it just counts any of the four possible hook calls ^^

**Review by awaelchli — APPROVED** [review]

File: tests/accelerators/test_cpu.py (hunk 2 above)
Comment (SeanNaren) [reply tests/accelerators/test_cpu.py:105]: I'll make the test slightly better by increasing granularity!

Comment (SeanNaren) [reply tests/accelerators/test_cpu.py]: It's still not perfect, but now atleast we have a per stage count to ensure we're more explicit

**Review by justusschock — APPROVED** [review]

**Review by carmocca — APPROVED** [review]
